ui_test/src/rustc_stderr.rs

```diff
@@ -0,0 +1,144 @@
+use std::{
+    fmt::Write,
+    path::{Path, PathBuf},
+};
+
+use regex::Regex;
+
+#[derive(serde::Deserialize, Debug)]
+struct RustcMessage {
+    rendered: Option<String>,
+    spans: Vec<Span>,
+    level: String,
+    message: String,
+    children: Vec<RustcMessage>,
+}
+
+#[derive(Copy, Clone, Debug, PartialOrd, Ord, PartialEq, Eq)]
+pub(crate) enum Level {
+    Error = 0,
+    Warn = 1,
+    Help = 2,
+    Note = 3,
+    /// Only used for "For more information about this error, try `rustc --explain EXXXX`".
+    FailureNote = 4,
+}
+
+impl std::str::FromStr for Level {
+    type Err = String;
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match s {
+            "ERROR" | "error" => Ok(Self::Error),
+            "WARN" | "warning" => Ok(Self::Warn),
+            "HELP" | "help" => Ok(Self::Help),
+            "NOTE" | "note" => Ok(Self::Note),
+            "failure-note" => Ok(Self::FailureNote),
+            _ => Err(format!("unknown level `{s}`")),
+        }
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct Message {
+    pub(crate) level: Level,
+    pub(crate) message: String,
+}
+
+impl RustcMessage {
+    fn line(&self, file: &Path) -> Option<usize> {
+        self.spans.iter().find_map(|span| span.line(file))
+    }
+
+    /// Put the message and its children into the line-indexed list.
+    fn insert_recursive(
+        self,
+        file: &Path,
+        messages: &mut Vec<Vec<Message>>,
+        messages_from_unknown_file_or_line: &mut Vec<Message>,
+        line: Option<usize>,
+    ) {
+        let line = self.line(file).or(line);
+        let msg = Message { level: self.level.parse().unwrap(), message: self.message };
+        if let Some(line) = line {
+            if messages.len() <= line {
+                messages.resize_with(line + 1, Vec::new);
+            }
+            messages[line].push(msg);
+        } else if !msg.message.starts_with("aborting due to") {
+            messages_from_unknown_file_or_line.push(msg);
+        }
+        for child in self.children {
+            child.insert_recursive(file, messages, messages_from_unknown_file_or_line, line)
+        }
+    }
+}
+
+impl Span {
+    fn line(&self, file: &Path) -> Option<usize> {
+        match &self.expansion {
+            None => (self.file_name == file).then_some(self.line_start),
+            Some(expansion) => expansion.span.line(file),
```

Comment:
So the "expansion" takes precedence, if it is set? Is that something about macros?

Reply:
yea, this is just about macro expansions. I guess we could try the current span first. This is relevant for when there are macros used and defined in the test itself.

Reply:
Trying multiple spans makes things a lot more complicated.^^

Seems fine to pick something and tweak our heuristic if we ever need to match an error in a macro. But please add a doc comment explaining that, and also document at the `expansion` field that this is about macro expansion.